test/integration/ferc1_eia_train_test.py

```diff
@@ -0,0 +1,116 @@
+"""Tests for the validation functions in the ferc1_eia_train module."""
+
+import importlib.resources
+import logging
+
+import pandas as pd
+import pytest
+
+from pudl.analysis.ferc1_eia import restrict_train_connections_on_date_range
+from pudl.analysis.ferc1_eia_train import validate_override_fixes
+from pudl.output.pudltabl import PudlTabl
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture(scope="module")
+def pudl_out(pudl_engine, pudl_datastore_fixture):
+    """A PUDL output object for use in CI."""
+    return PudlTabl(
+        pudl_engine,
+        ds=pudl_datastore_fixture,
+        freq="AS",
+        fill_tech_desc=False,
+        fill_net_gen=False,
+    )
+
+
+@pytest.fixture(scope="module")
+def utils_eia860(pudl_out):
+    """The utils_eia860 output table."""
+    return pudl_out.utils_eia860()
+
+
+@pytest.fixture(scope="module")
+def plant_part_list(pudl_out):
+    """The plant_parts_eia output table."""
+    return pudl_out.plant_parts_eia().reset_index()
+
+
+@pytest.fixture(scope="module")
+def ferc1_eia(pudl_out):
+    """The ferc1_eia output table."""
+    return pudl_out.ferc1_eia()
+
+
+@pytest.fixture(scope="module")
+def ferc1_eia_training_data():
+    """The training data for the ferc1_eia matching."""
+    return pd.read_csv(
+        importlib.resources.path("pudl.package_data.glue", "ferc1_eia_train.csv")
+    )
+
+
+@pytest.mark.parametrize(
+    "verified,report_year,record_id_eia_override_1,record_id_ferc1",
+    [
+        pytest.param(
+            ["True"],
+            [2020],
+            ["4270_2020_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3"],
+        ),
+        # This param fails for the wrong reason -- because the 2019 data isn't available.
+        pytest.param(
+            ["True"],
+            [2020],
+            ["4270_2019_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3"],
+            marks=pytest.mark.xfail(
+                reason="EIA record year doesn't match FERC record year"
+            ),
+        ),
+        pytest.param(
+            ["True", "True"],
+            [2020, 2020],
+            ["4270_2020_plant_total_11241", "4270_2020_plant_total_11241"],
+            ["f1_steam_2020_12_454_0_3", "f1_steam_2020_12_159_0_3"],
+            marks=pytest.mark.xfail(reason="Duplicate EIA ids in training data"),
```

Comment:
Made these xfail params, but they aren't guaranteed to fail for the right reasons. **Is there a way to ensure that they fail in a specific way?** The function this tests (`pudl.analysis.ferc1_eia_train.py::validate_override_fixes()`) throws a variety of AssertionErrors depending on whether certain conditionals are True or not. EX: 

```

    # Make sure there are no duplicate EIA id overrides
    logger.debug("Checking for duplicate override ids")
    assert (
        len(
            override_dups := only_overrides[
                only_overrides["record_id_eia_override_1"].duplicated(keep=False)
            ]
        )
        == 0
    ), f"Found record_id_eia_override_1 duplicates: \
    {override_dups.record_id_eia_override_1.unique()}"

    # Make sure the year in the EIA id overrides match the year in the report_year
    # column.
    logger.debug("Checking that year in override id matches report year")
    only_overrides = only_overrides.merge(
        ppl[["record_id_eia", "report_year"]].drop_duplicates(),
        left_on="record_id_eia_override_1",
        right_on="record_id_eia",
        how="left",
        suffixes=("", "_ppl"),
    )
    if (
        len(
            bad_eia_year := only_overrides["report_year"].compare(
                only_overrides["report_year_ppl"]
            )
        )
        > 0
    ):
        raise AssertionError(
            f"Found record_id_eia_override_1 values that don't correspond to the right \
            report year:\
            {[only_overrides.iloc[x].record_id_eia_override_1 for x in bad_eia_year.index]}"
        )
```

Reply:
I believe you can add a specific [type of assertion into the xfail parameters](https://doc.pytest.org/en/latest/how-to/skipping.html#raises-parameter).


So in this case it would be something like 
```
marks=pytest.mark.xfail(reason="Duplicate EIA ids in training data",  raises=AssertionError)
```

(i think this would necessitate you converting the bare assert into a more specific assert, which is best practice anyway)

but i don't think you can tell it specifically which possible error in a specific module that it should trip on. I think that would require breaking out these defensive tests into smaller functions and running them individually. Which I believe is a good overall design because it makes things easier to test. But if i understand correctly, you want to be able to run specific training examples through the full validation process.

One thing to note: I *believe* some of these integration tests are run during the CI process as well at during the full nightly builds so this xfail may not always hold. I'll go look into that a little bit bc I'm not sure.

Reply:
If you wanted to get super specific about why these tests are failing and when I think it would be much much easier to do if you broke out your `validate_override_fixes` into smaller chunks and tested each assert individually.